loopy/kernel/dependency.py

```diff
@@ -0,0 +1,227 @@
+    program.
+    """
+
+    # we want to modify the output dimension and OUT = 3
+    dim_type = isl.dim_type.out
+
+    # generate an unordered mapping from statement instances to points in the
+    # loop domain
+    insn_number = 0
+    schedules = {}
+    for insn in knl.instructions:
+        domain = knl.get_inames_domain(insn.within_inames)
+
+        # if we do not set the dim name, the name is set as None
+        domain = domain.insert_dims(dim_type, 0, 1).set_dim_name(dim_type, 0,
+                                                                 insn.id)
+
+        space = domain.get_space()
+        domain = domain.add_constraint(
+            isl.Constraint.eq_from_names(space, {1: -1*insn_number, insn.id: 1})
+        )
+
+        # this may not be the final way we keep track of the schedules
+        schedule = isl.Map.from_domain_and_range(domain, domain)
+        schedules[insn.id] = schedule
+
+        insn_number += 1
+
+        # determine a lexicographic order on the space the schedules belong to
+
+
+@for_each_kernel
+def add_lexicographic_happens_after(knl: LoopKernel) -> LoopKernel:
+
+    new_insns = []
+
+    for iafter, insn_after in enumerate(knl.instructions):
+        if iafter == 0:
+            new_insns.append(insn_after)
+        else:
+            insn_before = knl.instructions[iafter - 1]
+            shared_inames = insn_after.within_inames & insn_before.within_inames
+            unshared_before = insn_before.within_inames
+
+            domain_before = knl.get_inames_domain(insn_before.within_inames)
+            domain_after = knl.get_inames_domain(insn_after.within_inames)
+
+            happens_before = isl.Map.from_domain_and_range(
+                    domain_before, domain_after)
+            for idim in range(happens_before.dim(dim_type.out)):
+                happens_before = happens_before.set_dim_name(
+                        dim_type.out, idim,
+                        happens_before.get_dim_name(dim_type.out, idim) + "'")
+            n_inames_before = happens_before.dim(dim_type.in_)
+            happens_before_set = happens_before.move_dims(
+                    dim_type.out, 0,
+                    dim_type.in_, 0,
+                    n_inames_before).range()
+
+            shared_inames_order_before = [
+                    domain_before.get_dim_name(dim_type.out, idim)
+                    for idim in range(domain_before.dim(dim_type.out))
+                    if domain_before.get_dim_name(dim_type.out, idim)
+                    in shared_inames]
+            shared_inames_order_after = [
+                    domain_after.get_dim_name(dim_type.out, idim)
+                    for idim in range(domain_after.dim(dim_type.out))
+                    if domain_after.get_dim_name(dim_type.out, idim)
+                    in shared_inames]
+
+            assert shared_inames_order_after == shared_inames_order_before
+            shared_inames_order = shared_inames_order_after
```

Comment:
There's a big unresolved question here, in terms of what nesting order we should use for the shared inames. Right now, this uses the axis order in the domains, however we *also* already do have `LoopKernel.loop_priority` to indicate nesting order during code generation). If nothing else, this should make sure that the order produced is consistent with that. But there's also the option of using/introducing a different mechanism entirely for this.

@kaushikcfd, got an opinion?